### PR TITLE
Upgrade react-select to 2.0

### PR DESCRIPTION
**Pull request by zarsky-broad**
~Still needs a look at the multi-select from Lou.~

When the next version of react-select comes out, it'll support themes, which will allow us to simplify the style overrides a lot.

### CODE REVIEW

File: src/components/common.js [hunk 1]
```diff
@@ -199,8 +199,35 @@ export const comingSoon = span({
...
+        '&:hover': { borderColor: isFocused ? colors.blue[0] : undefined }
+      }),
+      singleValue: base => ({ ...base, color: colors.gray[0] }),
+      option: (base, { isSelected, isFocused, isDisabled }) => _.merge(base, {
+        backgroundColor: isSelected ? colors.blue[4] : isFocused ? colors.blue[5] : undefined,
+        color: isDisabled ? undefined : colors.gray[0],
+        ':active': { backgroundColor: isSelected ? colors.blue[4] : colors.blue[5] }
+      })
+    },
+    getOptionLabel: ({ value, label }) => label || _.flow(_.camelCase, _.startCase)(value), // start case doesn't handle all caps
```
Comment (panentheos) [inline src/components/common.js]: This seems a bit too clever/magical to put in the baseline component. In fact, there's a display bug in the 'memory' dropdown in the cluster manager when 1 cpu is selected (it mangles the decimal number). I think it's fine to support an array of scalars to make it easy when you want the same label and value, but when actual formatting is required, I think we should just be explicit and let the client specify it.

File: src/components/common.js [hunk 2]
```diff
@@ -199,8 +199,35 @@ export const comingSoon = span({
   }
 }, ['coming soon'])
 
-export const Select = RSelect
+export const Select = ({ value, options, ...props }) => {
+  const newOptions = options && !_.isObject(options[0]) ? _.map(value => ({ value }), options) : options
+  const findValue = target => _.find({ value: target }, newOptions)
+  const newValue = Utils.cond(
+    [_.isUndefined(value), undefined],
+    [_.has('value', value), value],
```
Comment (panentheos) [inline src/components/common.js]: I'd prefer to explicitly _not_ support this case. At best, I think this pattern encourages bad state management, and at worst could result in some really subtle bugs. We should always be specifying the underlying data, not the option object.

File: src/components/notebook-utils.js [hunk 3]
```diff
@@ -88,7 +88,7 @@ export const NotebookCreator = ajaxCaller(class NotebookCreator extends Componen
         tooltip: Utils.summarizeErrors(errors),
         onClick: () => {
           this.setState({ creating: true })
-          Buckets.notebook(namespace, bucketName, notebookName).create(notebookKernel.data).then(
+          Buckets.notebook(namespace, bucketName, notebookName).create(notebookData[notebookKernel]).then(
```
Comment (panentheos) [inline src/components/notebook-utils.js:91]: Yes, I think this is a better way of handling this.

File: src/pages/workspaces/workspace/ShareWorkspaceModal.js [hunk 4]
```diff
@@ -148,18 +149,18 @@ export default ajaxCaller(class ShareWorkspaceModal extends Component {
...
-            wrapperStyle: styles.roleSelect,
-            searchable: false, clearable: false,
-            disabled: isMe,
+            styles: { container: styles.roleSelect },
+            isSearchable: false, isClearable: false,
+            isDisabled: isMe,
             value: accessLevel,
-            onChange: ({ value }) => this.setState({ acl: _.set([index, 'accessLevel'], value, acl) }),
-            options: _.map(level => ({ label: level, value: level.toUpperCase() }), ['Reader', 'Writer', 'Owner'])
+            onChange: ({ value }) => this.setState({ acl: _.set([index, 'accessLevel'], value.toUpperCase, acl) }),
```
Comment (panentheos) [inline src/pages/workspaces/workspace/ShareWorkspaceModal.js]: Should this be `toUpperCase()`?

File: src/components/NewWorkspaceModal.js [hunk 5]
```diff
@@ -155,18 +153,20 @@ export default ajaxCaller(class NewWorkspaceModal extends Component {
...
       ])),
       !!existingGroups.length && div({ style: styles.groupNotice }, [
-        'The cloned workspace will automatically inherit the authorization domain from this workspace. ',
-        'You may add groups to the authorization domain, but you may not remove existing ones.'
+        div({ style: { marginBottom: '0.2rem' } }, [
+          'The cloned workspace will automatically inherit the authorization domain from this workspace. ',
+          'You may add groups to the authorization domain, but you may not remove existing ones.'
+        ]),
+        div({ style: { color: colors.gray[0], marginBottom: '0.2rem' } }, ['Inherited groups:']),
+        div({ style: { color: colors.gray[0] } }, existingGroups.join(', '))
```
Comment (panentheos) [inline src/components/NewWorkspaceModal.js]: I take it the new version doesn't support individual clearable flags on the options?

File: src/components/ClusterManager.js [hunk 6]
```diff
@@ -325,20 +326,26 @@ export default ajaxCaller(class ClusterManager extends PureComponent {
...
             value: profile,
+            getOptionLabel: ({ value }) => {
+              if (value === 'custom') {
+                return 'Custom'
+              } else {
+                const { label, machineConfig } = profilesByName[value]
+
+                return `${label} computer power (${Utils.formatUSD(machineConfigCost(machineConfig))} hr)`
+              }
+            },
```
Comment (panentheos) [inline src/components/ClusterManager.js]: The old code would have worked without modification, right? Why do we need to use `getOptionLabel` here?

Comment (zarsky-broad) [reply src/components/ClusterManager.js]: Mainly, just thinking about establishing a new pattern, I did this file first... I'll back this one out.

File: src/components/NewWorkspaceModal.js (hunk 5 above)
Comment (zarsky-broad) [reply src/components/NewWorkspaceModal.js]: Correct.

File: src/components/notebook-utils.js [hunk 7]
```diff
@@ -109,30 +109,14 @@ export const NotebookCreator = ajaxCaller(class NotebookCreator extends Componen
           onChange: e => this.setState({ notebookName: e.target.value, nameTouched: true })
         }
       }),
-      Forms.requiredFormLabel('Kernel'),
+      Forms.requiredFormLabel('Language'),
       h(Select, {
-        clearable: false,
-        searchable: false,
+        isSearchable: false,
         placeholder: 'Select a kernel',
```
Comment (MattPutnam) [inline src/components/notebook-utils.js]: Another kernel

**Review by MattPutnam — COMMENTED** [review]
I don't like how the first item looks like it's being hovered when you open the menu with a click. I also still don't like the behavior in the share modal, though it's better than it was.

File: src/components/NewWorkspaceModal.js [hunk 8]
```diff
@@ -79,7 +78,7 @@ export default ajaxCaller(class NewWorkspaceModal extends Component {
       const body = {
         namespace,
         name,
-        authorizationDomain: _.map(v => ({ membersGroupName: v }), groups),
+        authorizationDomain: [..._.map(v => ({ membersGroupName: v }), groups), ...cloneWorkspace.workspace.authorizationDomain],
```
Comment (panentheos) [inline src/components/NewWorkspaceModal.js]: Need to check for undefined here. This fails if you try to create a workspace from scratch.

**Review by panentheos — APPROVED** [review]
Looks good, but see bug comment.

**Comment by katevoss** [comment]
@zarsky-broad can you provide any information about how you recommend testing this? I don't know what to look for.

**Comment by zarsky-broad** [comment]
@katevoss Every instance of react-select, that is, every dropdown in the application, should work as expected after this upgrade. Take a look a the places that are changed in code to find them.

File: src/components/NewWorkspaceModal.js (hunk 8 above)
Comment (zarsky-broad) [reply src/components/NewWorkspaceModal.js]: I can't reproduce this.

Comment (zarsky-broad) [reply src/components/NewWorkspaceModal.js]: Oop, I can't reproduce this on my local, which I've already patched but not pushed, heheh.

**Comment by zarsky-broad** [comment]
@MattPutnam We can potentially fix those things in a follow-on. For now I'm going ahead with this.

**Pull request merged**